lib/web_ui/lib/src/engine/view_embedder/dom_manager.dart

```diff
@@ -19,24 +22,111 @@ import '../embedder.dart';
-///        |
-///        +- ...semantics nodes
+///   +- [semanticsHost] <flt-semantics-host>
+///   |    |
+///   |    +- ...semantics nodes
+///   |
+///   +- <style>
 ///
 class DomManager {
-  DomManager.fromFlutterViewEmbedderDEPRECATED(this._embedder);
+  factory DomManager({required double devicePixelRatio}) {
+    final DomElement rootElement = domDocument.createElement(DomManager.flutterViewTagName);
+    final DomElement platformViewsHost = domDocument.createElement(DomManager.glassPaneTagName);
+    final DomShadowRoot renderingHost = _attachShadowRoot(platformViewsHost);
+    final DomElement sceneHost = domDocument.createElement(DomManager.sceneHostTagName);
+    final DomElement textEditingHost = domDocument.createElement(DomManager.textEditingHostTagName);
+    final DomElement semanticsHost = domDocument.createElement(DomManager.semanticsHostTagName);
+    final DomElement announcementsHost = createDomElement(DomManager.announcementsHostTagName);
+
+    // Root element children.
+
+    rootElement.appendChild(platformViewsHost);
+    rootElement.appendChild(textEditingHost);
+    // The semantic host goes last because hit-test order-wise it must be
+    // first. If semantics goes under the scene host, platform views will
+    // obscure semantic elements.
+    //
+    // You may be wondering: wouldn't semantics obscure platform views and
+    // make then not accessible? At least with some careful planning, that
+    // should not be the case. The semantics tree makes all of its non-leaf
+    // elements transparent. This way, if a platform view appears among other
+    // interactive Flutter widgets, as long as those widgets do not intersect
+    // with the platform view, the platform view will be reachable.
+    rootElement.appendChild(semanticsHost);
+
+    // Rendering host (shadow root) children.
+
+    // TODO(yjbanov): In a multi-view world, we may want to have a separate
+    //                semantics owner for each view.
+    //                https://github.com/flutter/flutter/issues/137445
+    final DomElement accessibilityPlaceholder = EngineSemanticsOwner
+        .instance.semanticsHelper
+        .prepareAccessibilityPlaceholder();
+
+    renderingHost.append(accessibilityPlaceholder);
+    renderingHost.append(sceneHost);
+    renderingHost.append(announcementsHost);
+
+    // Styling.
+
+    StyleManager.attachGlobalStyles(
+      node: rootElement,
+      styleId: 'flt-text-editing-stylesheet',
+      styleNonce: configuration.nonce,
+      cssSelectorPrefix: DomManager.flutterViewTagName,
+    );
+
+    StyleManager.attachGlobalStyles(
+      node: renderingHost,
+      styleId: 'flt-internals-stylesheet',
+      styleNonce: configuration.nonce,
+      cssSelectorPrefix: '',
+    );
+
+    StyleManager.styleSceneHost(
+      sceneHost,
+      debugShowSemanticsNodes: configuration.debugShowSemanticsNodes,
+    );
+
+    StyleManager.styleSemanticsHost(
+      semanticsHost,
+      devicePixelRatio,
```

Comment:
Looks like `devicePixelRatio` is only used here, and also it seems like a bug? How do we update the semantics host when `devicePixelRatio` changes?

Reply:
> Looks like `devicePixelRatio` is only used here

Not sure what you mean here. Yes, it's not used anywhere else in this method.

> and also it seems like a bug? How do we update the semantics host when `devicePixelRatio` changes?

The code that listens for resize events (and updates the semantics host) is still in `FlutterViewEmbedder` for now:
https://github.com/flutter/engine/blob/aae07e989b0a031c1b1fd2739cd10757e7cbd67f/lib/web_ui/lib/src/engine/embedder.dart#L220-L224

I do plan to move it out of there though. Stay tuned.

Reply:
I'm just having a feeling that this is the wrong place for styling and scaling the semantics host. In fact, I'm thinking maybe the semantics host is the wrong element to style and scale. Should it be the responsibility of `EngineSemanticsOwner`? Then we won't need `DomManager` to be concerned about `devicePixelRatio`.

Having said that, I'm totally fine with not making this change in this PR.

Reply:
Yeah, I think it makes sense for the semantic owner to "own" the styling of the semantics tree.

I'll leave it as is for now, but feel free to move it to the semantics owner as part of your overhaul there.